### PR TITLE
feat(CodeEditor): use new monaco package

### LINKED ISSUE #8578: React Code Editor - Replace `react-monaco-editor` with `@monaco-editor/react`
The [`react-monaco-editor`](https://github.com/react-monaco-editor/react-monaco-editor) package should be replaced with [`@monaco-editor/react`](https://github.com/suren-atoyan/monaco-react). The latter is much more mature and receives more frequent updates, specifically:

- It has much better documentation
- Doesn't need a WebPack plugin
- Has a stable version

I think this would be a good candidate for the next major version of PatternFly React.

### PR DESCRIPTION
**Pull request by kmcfaul**
**What**: Closes #8578

Will put up an org PR as well updating the dependency on that end (it's looking like this PR will have to merge first to update the versions on that end to build successfully). 

There doesn't seem to be much difference between the two packages in how its being consumed, mostly some type definitions are not exported in the newer package so I exported those types in the component manually.

Added the package as a dev dependency in order to get it building locally automatically.

### CODE REVIEW

**Comment by kmcfaul** [comment]
@nicolethoen One open question I've got for this PR is if we're okay with merging this into the current package or if we need to move some stuff around. The only really outward change should be the slight update to the `EditorDidMount` type (since the new package doesn't export the types directly), and the package name / dependency.

**Comment by nicolethoen** [comment]
@kmcfaul 
we should make sure there are clear follow up issues marked as next steps as part of this whole story being complete.
Since it sounds like you will need to update the webpack config in the documentation framework, then do a version bump in pf-react to update the docs-framework dependency and remove the new peer dependency.

I think this is a breaking change, so it will have to be introduced as a code-editor-next, which I think might be easiest to do in a new package, but that can be up to you and @tlabaj. It's breaking change because of the follow up work all consumers will want/need to do to clean up their webpack configurations and the `EditorDidMount` prop name update 
If it's not going in it's own package, you'll have to extensively update the README to be clear when the webpack configurations are needed and when they are not.

File: packages/react-code-editor/package.json [hunk 1]
```diff
@@ -38,10 +38,11 @@
   "peerDependencies": {
     "react": "^17 || ^18",
     "react-dom": "^17 || ^18",
-    "react-monaco-editor": "^0.51.0"
```
Comment (tlabaj) [inline packages/react-code-editor/package.json:41]: I believe changing the Peer dependency here is breaking.  We would want to introduce this as next.

**Comment by kmcfaul** [comment]
Opened https://github.com/patternfly/patternfly/issues/6115 as a follow up to update the readme.

Per conversation, moved the new package to a direct dependency.

**Review by tlabaj — APPROVED** [review]
LGTM!

**Review by andrew-ronaldson — APPROVED** [review]
Approved! Thanks

**Review by thatblindgeye — APPROVED** [review]
CSS/HTMl wise looks good. Only non-blocker I noticed is the "With shortcut menu and main header content" example doesn't actually have any header content passed in, but that's more an example/docs issue.

**Comment by jonkoops** [comment]
Just FYI, there are still some references to `react-monaco-editor` in the [Jest config](https://github.com/patternfly/patternfly-react/blob/c90c7f26330e3d94ae5927318060b6d2a7faeef6/jest.config.js#L17) and [README](https://github.com/patternfly/patternfly-react/blob/c90c7f26330e3d94ae5927318060b6d2a7faeef6/packages/react-code-editor/README.md?plain=1#L52) that might have to be updated. The README still seems to reference some other things that are no longer relevant, such as the WebPack plugin.

**Pull request merged**